### ISSUE TITLE
🚀🚀 Model Compression and Model Acceleration in TensorLayer

### ISSUE THREAD

**Issue by zsdonghao** [tensorlayer/tensorlayer#416]
http://machinethink.net/blog/compressing-deep-neural-nets/

## ------- before 12 Aug 2018 -------

 - SqueezeNet, see [tutorial_squeezenet.py](https://github.com/zsdonghao/tensorlayer/blob/master/example/tutorial_squeezenet.py)
- MobileNet, see [tutorial_mobilenet.py](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_mobilenet.py)
 - BinaryNet, see [mnist](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_binarynet_mnist_cnn.py) [cifar10](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_binarynet_cifar10_tfrecord.py).
 - Ternary Weight Network, see [mnist](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_ternaryweight_mnist_cnn.py) [cifar10](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_ternaryweight_cifar10_tfrecord.py).
 - DoReFa-Net, see [mnist](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_dorefanet_mnist_cnn.py) [cifar10](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_dorefanet_cifar10_tfrecord.py).

@XJTUWYD add two compress strategies tenary weight network  and dorefa-net into tensorlayer,  did two experiments to compare the accuracy of different compress strategies based on mnist and cifar-10.
the result of the experiment is below:

|             	| BinaryNet      	| Tenary Weight      	| DoReFa-Net      	|
|-------------	|---------------	|---------------	|---------------	|
| MNIST 	| 98.86% 	| 99.27% 	| 98.89% 	|
| CIFAR10  	| 41.1% 	| 80.6% 	| 81.1% 	|

@XJTUWYD : bnn is  a excellent work in the compression of neuron network but it can not get a satisfied accuracy on relative large datasets, in order to solve the problem, tenary weight networks and dorefa were put forward. I add 4 apis for tensorlayer, Tenary Denselayer, TenaryConv2d, DorefaDenselayer, and DorefaConv2d . I perform 6 experiment based on mnist and cifar10，the details are in thr tutorials. Finally thank you very much for the help of HaoDong, LuoMai, and Igarithm.

## ------- before 15 March 2018 -------

Hi, I am trying to make TensorLayer/ TensorFlow to support  BinaryNet, Xnor-Net, SqueezeNet, MobileNet, ShuffleNet, DoReFaNet, Channel Pruning and etc. Feel free to discuss and add more information here. ~

### Paper List

- [A Survey of Model Compression and Acceleration for Deep Neural Networks](https://arxiv.org/pdf/1710.09282v5.pdf) (end of 2017)
- More in [arXiv-Sanity](http://www.arxiv-sanity.com/search?q=model+compression)

#### 1. Quantization

- Quantized Neural Networks: Training Neural Networks with Low Precision Weights and Activations [paper](https://arxiv.org/pdf/1609.07061.pdf)
- Binarized Neural Networks: Training Deep Neural Networks with Weights and Activations Constrained to +1 or -1 [paper](https://arxiv.org/abs/1602.02830) [itayhubara/BinaryNet.tf](https://github.com/itayhubara/BinaryNet.tf)
- Ternary Weight Networks [paper](https://arxiv.org/abs/1605.04711)
- XNOR-Net: ImageNet Classification Using Binary Convolutional Neural Networks [paper](https://arxiv.org/abs/1603.05279) [AngusG/tensorflow-xnor-bnn](https://github.com/AngusG/tensorflow-xnor-bnn/blob/master/models/binary_net.py#L79)
- DoReFa-Net: Training Low Bitwidth Convolutional Neural Networks with Low Bitwidth Gradients [paper](https://arxiv.org/abs/1606.06160)
- Towards Accurate Binary Convolutional Neural Network [paper](https://arxiv.org/abs/1711.11294)

#### 2. Pruning

- Channel Pruning for Accelerating Very Deep Neural Networks [paper](https://arxiv.org/abs/1707.06168)

#### 3. Structure

- SqueezeNet: AlexNet-level accuracy with 50x fewer parameters and <0.5MB model size [paper](https://arxiv.org/abs/1602.07360) [TensorLayer example](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_squeezenet.py)
- MobileNets: Efficient Convolutional Neural Networks for Mobile Vision Applications [paper](https://arxiv.org/abs/1704.04861)
- ShuffleNet: An Extremely Efficient Convolutional Neural Network for Mobile Devices [paper](https://arxiv.org/abs/1707.01083) [code1](https://github.com/ppwwyyxx/tensorpack/tree/master/examples/ImageNetModels) [code2](https://github.com/scheckmedia/keras-shufflenet/blob/master/shufflenet.py)
- [纵览轻量化卷积神经网络：SqueezeNet、MobileNet、ShuffleNet、Xception](https://www.jiqizhixin.com/articles/2018-01-08-6)
- [CNN 模型压缩与加速算法综述](https://cloud.tencent.com/developer/article/1005738)

### About TensorLayer & TensorFlow
- [tf.custom_gradient](https://www.tensorflow.org/versions/master/api_docs/python/tf/custom_gradient?hl=ES)
- [Quantized Layers](http://tensorlayer.readthedocs.io/en/latest/modules/layers.html)

### Others
- [How to Quantize Neural Networks with TensorFlow](https://www.tensorflow.org/performance/quantization)
- [how-to-quantize-neural-networks-with-tensorflow 2016 May](https://petewarden.com/2016/05/03/how-to-quantize-neural-networks-with-tensorflow/)
- [Embedded Vision Summit](https://www.embedded-vision.com/summit)
- [Deep Compression and EIE](http://on-demand.gputechconf.com/gtc/2016/presentation/s6561-song-han-deep-compression.pdf)
- [EIE: Efficient Inference Engine on Compressed Deep Neural Network](https://arxiv.org/pdf/1602.01528.pdf)

**Comment by zsdonghao** [comment]
https://github.com/tensorlayer/tensorlayer/releases/tag/1.8.2

This is an experimental API package for building Binary Nets. We are using matrix multiplication rather than add-minus and bit-count operation at the moment. Therefore, these APIs would not speed up the inferencing, for production, you can train model via TensorLayer and deploy the model into other customized C/C++ implementation (We probably provide users an extra C/C++ binary net framework that can load model from TensorLayer).

### TODO
- For binary input, use XNOR, bit-count operation to replace matrix multiplication.
- For non binary input, use add and minus operation to replace matrix multiplication.

**Comment by zsdonghao** [comment]
### Release SqueezeNet Example
https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_squeezenet.py

**Comment by zsdonghao** [comment]
## Release DoReFa-Net, BinaryNet, Tenary Weight Network 🚀

 - SqueezeNet (ImageNet). Classification task, see [tutorial_squeezenet.py](https://github.com/zsdonghao/tensorlayer/blob/master/example/tutorial_squeezenet.py)
 - BinaryNet. Model compression, see [mnist](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_binarynet_mnist_cnn.py) [cifar10](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_binarynet_cifar10_tfrecord.py).
 - Tenary Weight Network. Model compression, see [mnist](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_tenaryweight_mnist_cnn.py) [cifar10](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_tenaryweight_cifar10_tfrecord.py).
 - DoReFa-Net. Model compression, see [mnist](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_dorefanet_mnist_cnn.py) [cifar10](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_dorefanet_cifar10_tfrecord.py).

**Comment by zsdonghao** [comment]
### Release MobileNet
https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_mobilenet.py

- Note that, TensorLayer MobileNet is 3 time faster than Keras (same weights and architecture).
- TensorLayer takes 0.001~0.002 second for one image.
- Keras takes 0.004~0.005 second for one image.

**Comment by zsdonghao** [comment]
@filipetrocadoferreira check the weights here https://github.com/tensorlayer/pretrained-models

**Comment by filipetrocadoferreira** [comment]
I'm using `mobilenet.npz`

**Comment by ABILIBILI** [comment]
I wonder why the Tenary Weight Net name in the release version `TensorLayer 1.8.3` is incompatible with the branch master, `TenaryDenseLayer`, `TenaryConv2d` for `TensorLayer 1.8.3`,however `TernaryDenseLayer`,`TernaryConv2d` for master.So that if I use the release version,source code (`tutorial_ternaryweight_mnist_cnn.py` for example) refer to these class wouldn't work and I have to change the releated name.

**Comment by ABILIBILI** [comment]
I wonder why the Tenary Weight Net name in the release version `TensorLayer 1.8.3` is incompatible with the branch master, `TenaryDenseLayer`, `TenaryConv2d` for `TensorLayer 1.8.3`,however `TernaryDenseLayer`,`TernaryConv2d` for master.So that if I use the release version,source code (`tutorial_ternaryweight_mnist_cnn.py` for example) refer to these class wouldn't work and I have to change the releated name.
